src/Dictionaries/YAMLRegExpTreeDictionarySource.cpp

```diff
@@ -0,0 +1,350 @@
+#include <DataTypes/DataTypesNumber.h>
+#include <DataTypes/Serializations/ISerialization.h>
+
+#include <Core/ColumnsWithTypeAndName.h>
+
+#include <Dictionaries/DictionarySourceFactory.h>
+#include <Dictionaries/DictionarySourceHelpers.h>
+#include <Dictionaries/DictionaryStructure.h>
+
+#include <Interpreters/Context.h>
+#include <Processors/Sources/SourceFromSingleChunk.h>
+
+#include <Poco/Util/AbstractConfiguration.h>
+
+#include <Common/filesystemHelpers.h>
+#include <Common/logger_useful.h>
+
+namespace DB
+{
+
+inline const String kYAMLRegExpTreeDictionarySource = "YAMLRegExpTreeDictionarySource";
+inline const String kYAMLRegExpTree = "yamlregexptree";
+
+namespace ErrorCodes
+{
+    extern const int LOGICAL_ERROR;
+    extern const int SUPPORT_IS_DISABLED;
+    extern const int INCORRECT_DICTIONARY_DEFINITION;
+}
+
+void registerDictionarySourceYAMLRegExpTree(DictionarySourceFactory & factory)
+{
+    auto create_table_source = [=]([[maybe_unused]] const DictionaryStructure & dict_struct,
+                                   [[maybe_unused]] const Poco::Util::AbstractConfiguration & config,
+                                   [[maybe_unused]] const String & config_prefix,
+                                   Block & ,
+                                   [[maybe_unused]] ContextPtr global_context,
+                                   const String &,
+                                   [[maybe_unused]] bool created_from_ddl) -> DictionarySourcePtr
+    {
+#if USE_YAML_CPP
+        if (dict_struct.has_expressions)
+        {
+            throw Exception(
+                ErrorCodes::LOGICAL_ERROR, "Dictionary source of type `{}` does not support attribute expressions", kYAMLRegExpTree);
+        }
+
+        if (!dict_struct.key.has_value() || dict_struct.key.value().size() != 1 || (*dict_struct.key)[0].type->getName() != "String")
+        {
+            throw Exception(ErrorCodes::INCORRECT_DICTIONARY_DEFINITION, "dictionary source `{}` should have one primary key with string value to represent regular expressions", kYAMLRegExpTree);
+        }
+
+        const auto & filepath = config.getString(config_prefix + "." + kYAMLRegExpTree + ".path");
+
+        const auto & context = copyContextAndApplySettingsFromDictionaryConfig(global_context, config, config_prefix);
+
+        return std::make_unique<YAMLRegExpTreeDictionarySource>(filepath, dict_struct, context, created_from_ddl);
+#else
+        throw Exception(
+            ErrorCodes::SUPPORT_IS_DISABLED,
+            "Dictionary source of type `{}` is disabled because ClickHouse was built without yaml_cpp support",
+            kYAMLRegExpTree);
+#endif
+    };
+
+    factory.registerSource(kYAMLRegExpTree, create_table_source);
+}
+
+}
+
+#if USE_YAML_CPP
+
+namespace DB
+
+{
+
+/**
+*   Configuration allowed fields
+*
+*   Example:
+*   ```
+*   - regexp: "MSIE (\\d+)"
+*     attr1: "Vecna"
+*     attr2: 22.8
+*     match: # nested match for subpattern
```

Comment:
in tests it's named `versions`, so what is the correct way to define subpath in this tree?

Reply:
indeed, the subpath attrubute name can be arbitrary.